File: src/algorithms/python/Kth_Largest_Element_in_an_Array.py
# -*- coding: utf-8 -*-

import heapq


class Solution:
    # @param {integer[]} nums
    # @param {integer} k
    # @return {integer}
    def findKthLargest(self, nums, k):
        heap = []
        for num in nums:
            if len(heap) < k:
                heapq.heappush(heap, num)
            else:
                if heap[0] < num:
                    heapq.heappop(heap)
                    heapq.heappush(heap, num)
        return heapq.heappop(heap)


if __name__ == "__main__":
    solution = Solution()
    print solution.findKthLargest([3, 2, 1, 5, 6, 4], 2)

File: src/algorithms/cpp/Kth_Largest_Element_in_an_Array.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <queue>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue <int, std::vector<int>, std::greater<int>> q;
        for (auto &num: nums) {
            if (SZ(q) < k)
                q.push(num);
            else {
                if (q.top() < num) {
                    q.pop();
                    q.push(num);
                }
            }
        }
        return q.top();
    }
};

int main() {
    vector <int> nums = {1, 2};
    Solution solution;
    cout << solution.findKthLargest(nums, 2) << endl;
    return 0;
}
